src/routing_table/mod.rs

```diff
@@ -922,93 +967,77 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
                 .iter()
                 .partition::<BTreeSet<_>, _>(|name| self.our_prefix.matches(name));
         self.our_section = our_new_section;
+        self.our_version = version + 1;
         // Drop sections that ceased to be our neighbours.
         let sections_to_remove = self.sections
             .keys()
             .filter(|prefix| !prefix.is_neighbour(&self.our_prefix))
             .cloned()
             .collect_vec();
-        self.insert_new_section(other_prefix, other_section);
+        self.insert_new_section(other_prefix, version + 1, other_section);
         sections_to_remove
             .into_iter()
-            .filter_map(|prefix| self.sections.remove(&prefix))
+            .filter_map(|prefix| {
+                            self.sections
+                                .remove(&prefix)
+                                .map(|(_, section)| section)
+                        })
             .flat_map(BTreeSet::into_iter)
             .collect()
     }
 
     /// Inserts the given section. Logs an error if it already exists.
-    fn insert_new_section(&mut self, prefix: Prefix<T>, section: BTreeSet<T>) {
+    fn insert_new_section(&mut self, prefix: Prefix<T>, version: u64, section: BTreeSet<T>) {
         match self.sections.entry(prefix) {
             Entry::Vacant(entry) => {
-                let _section_ref = entry.insert(section);
+                let _section_ref = entry.insert((version, section));
             }
             Entry::Occupied(entry) => {
                 error!("{:?} Inserting section {:?}, but already has members {:?}. This is a bug!",
                        self.our_name,
                        prefix,
                        entry.get());
-                entry.into_mut().extend(section);
+                let &mut (ref mut v, ref mut s) = entry.into_mut();
+                if *v > version {
```

Comment:
I'm not sure about this. I'd rather not rely on the current set of call sites.
And the problem with potentially re-adding a dropped peer is probably not new with this PR? I don't know what's the best solution here…